pkg/kubelet/lifecycle/predicate.go

```diff
@@ -153,11 +154,46 @@ func (w *predicateAdmitHandler) Admit(attrs *PodAdmitAttributes) PodAdmitResult
 			Message: message,
 		}
 	}
+	// By this time, node labels should have been synced, this helps in identifying the pod with the usage.
+	if rejectPodAdmissionBasedOnOSField(admitPod, node) {
+		return PodAdmitResult{
+			Admit:   false,
+			Reason:  "PodOSFieldNodeLabelDoesNotMatch",
+			Message: "Failed to admit pod as the OS field doesn't match node OS",
+		}
+	}
 	return PodAdmitResult{
 		Admit: true,
 	}
 }
 
+// rejectPodAdmissionBasedOnOSField rejects pods if their OS field doesn't match node's OS label.
+// TODO: Relax this restriction when we start supporting LCOW in kubernetes where podOS may not match
+// 		 node's OS.
+func rejectPodAdmissionBasedOnOSField(pod *v1.Pod, node *v1.Node) bool {
+	if pod.Spec.OS == nil {
+		return false
+	}
+	if !feature.DefaultFeatureGate.Enabled(features.IdentifyPodOS) {
+		return false
+	}
+	labels := node.Labels
+	osName, osLabelExists := labels[v1.LabelOSStable]
+	if !osLabelExists || osName != runtime.GOOS {
+		if len(labels) == 0 {
+			labels = make(map[string]string)
+		}
+		labels[v1.LabelOSStable] = runtime.GOOS
+	}
+	// Node OS label can be used as source of truth as the OS and arch reconciliation would have happened by now
+	if nodeOS, ok := labels[v1.LabelOSStable]; ok {
+		if string(pod.Spec.OS.Name) != nodeOS {
```

Comment:
don't compare the pod os to the node labels, compare it to runtime.GOOS

Reply:
the pod nodeSelector should be compared to the node labels with the OS/Arch label ensured to match the real runtime.GOOS/GOARCH, but that's distinct from checking the pod.spec.os against runtime.GOOS